discretize/base/base_mesh.py

```diff
@@ -2246,6 +2657,285 @@ def get_edge_inner_product_deriv(
+                        stacklevel=2,
+                    )
+                    return dMdprop
+                return sdiag(v) * dMdprop
+
+            return innerProductDeriv
+        else:
+            return None
+
+    def get_edge_inner_product_line_deriv(
+        self,
+        model,
+        invert_model=False,
+        invert_matrix=False,
+        **kwargs,
+    ):
+        r"""Get a function handle to multiply a vector with derivative of edge inner product line matrix (or its inverse).
+
+        Let :math:`\mathbf{M}(\mathbf{m})` be the edge inner product line matrix
+        constructed with a set of diagnostic property parameters :math:`\mathbf{m}`
+        (or its inverse) defined on mesh edges. **get_edge_inner_product_line_deriv**
+        constructs a function handle
+
+        .. math::
+            \mathbf{F}(\mathbf{u}) = \mathbf{u}^T \, \frac{\partial \mathbf{M}(\mathbf{m})}{\partial \mathbf{m}}
+
+        which accepts any numpy.array :math:`\mathbf{u}` of shape (n_edges,). That is,
+        **get_edge_inner_product_line_deriv** constructs a function handle for computing
+        the dot product between a vector :math:`\mathbf{u}` and the derivative of the
+        edge inner product line matrix (or its inverse) with respect to the diagnostic parameters.
+        When computed, :math:`\mathbf{F}(\mathbf{u})` returns a ``scipy.sparse.csr_matrix``
+        of shape (n_edges, n_param).
+
+        The function handle can only be created for isotropic diagnostic properties.
+
+        Parameters
+        ----------
+        model : (n_edges, ) numpy.ndarray
+            Parameters defining the diagnostic property values for every edge in the mesh.
+        invert_model : bool, optional
+            The inverse of *model* is used as the diagnostic property.
+        invert_matrix : bool, optional
+            Returns the inverse of the inner product line matrix.
+
+        Returns
+        -------
+        function
+            The function handle :math:`\mathbf{F}(\mathbf{u})` which accepts a
+            (``n_edges``) :class:`numpy.ndarray` :math:`\mathbf{u}`. The function
+            returns a (``n_edges``, ``n_params``) :class:`scipy.sparse.csr_matrix`.
+        """
+        try:
+            if model is None:
+                tensorType = -1
+            elif is_scalar(model):
+                tensorType = 0
+            elif model.size == self.nE:
+                tensorType = 1
+            else:
+                raise Exception(
+                    "Unexpected shape of tensor: {}.".format(model.shape),
+                    "Must be scalar or have length equal to total number of edges: {}.".format(
+                        self.nE
+                    ),
+                )
+
+            L = sdiag(self.edge_lengths)
+
+        except NotImplementedError:
+            raise NotImplementedError(
+                f"get_edge_inner_product_line_deriv not implemented for {type(self)}"
+            )
```

Comment:
Similar comments to above. Move the raised exceptions closer to what they are testing.